server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -261,18 +219,9 @@ private SimpleTopDocsCollectorContext(
                 topDocsSupplier = new CachedSupplier<>(topDocsCollector::topDocs);
                 totalHitsSupplier = () -> new TotalHits(0, TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
             } else {
-                // implicit total hit counts are valid only when there is no filter collector in the chain
-                final int hitCount = hasFilterCollector ? -1 : shortcutTotalHitCount(reader, query);
-                if (hitCount == -1) {
-                    topDocsCollector = createCollector(sortAndFormats, numHits, searchAfter, trackTotalHitsUpTo);
-                    topDocsSupplier = new CachedSupplier<>(topDocsCollector::topDocs);
-                    totalHitsSupplier = () -> topDocsSupplier.get().totalHits;
-                } else {
-                    // don't compute hit counts via the collector
-                    topDocsCollector = createCollector(sortAndFormats, numHits, searchAfter, 1);
```

Comment:
I like this change, this is the only place where I think we may be seeing regressions. Before your change, we would tell the top docs collector that it doesn't have to count hits at all (the `1` here) since we computed it up-front. With your change, we would always count `trackTotalHitsUpTo` documents, which delays a bit skipping.

Maybe we could run benchmarks with the `geonames` track to quantify the impact. I'll be especially interested on the impact on the `default` and `term` queries.